spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/RetryLoadBalancerInterceptor.java

```diff
@@ -69,27 +77,97 @@ public ClientHttpResponse intercept(final HttpRequest request, final byte[] body
-						}
-						ClientHttpResponse response = RetryLoadBalancerInterceptor.this.loadBalancer.execute(
-								serviceName, serviceInstance,
-								requestFactory.createRequest(request, body, execution));
-						int statusCode = response.getRawStatusCode();
-						if(retryPolicy != null && retryPolicy.retryableStatusCode(statusCode)) {
-							response.close();
-							throw new RetryableStatusCodeException(serviceName, statusCode);
-						}
-						return response;
-					}
-				});
+                    @Override
+                    public ClientHttpResponse doWithRetry(RetryContext context)
+                            throws IOException {
+                        ServiceInstance serviceInstance = null;
+                        if (context instanceof LoadBalancedRetryContext) {
+                            LoadBalancedRetryContext lbContext = (LoadBalancedRetryContext) context;
+                            serviceInstance = lbContext.getServiceInstance();
+                        }
+                        if (serviceInstance == null) {
+                            serviceInstance = loadBalancer.choose(serviceName);
+                        }
+                        ClientHttpResponse response = RetryLoadBalancerInterceptor.this.loadBalancer.execute(
+                                serviceName, serviceInstance,
+                                requestFactory.createRequest(request, body, execution));
+                        int statusCode = response.getRawStatusCode();
+                        if (retryPolicy != null && retryPolicy.retryableStatusCode(statusCode)) {
+                            ClientHttpResponseWrapper wrapper = new ClientHttpResponseWrapper(response);
+                            wrapper.init();
+                            throw new RetryableStatusCodeException(serviceName, statusCode, wrapper, null);
+                        }
+                        return response;
+                    }
+                }, new RecoveryCallback<ClientHttpResponse>() {
+                    @Override
+                    public ClientHttpResponse recover(RetryContext retryContext) throws Exception {
+                        Throwable lastThrowable = retryContext.getLastThrowable();
+                        if (lastThrowable != null && lastThrowable instanceof RetryableStatusCodeException) {
+                            RetryableStatusCodeException ex = (RetryableStatusCodeException) lastThrowable;
+                            return (ClientHttpResponse) ex.getResponse();
+                        }
+                        if (lastThrowable instanceof Error) {
+                            throw (Error) lastThrowable;
+                        }
+                        if (lastThrowable instanceof Exception) {
+                            throw (Exception) lastThrowable;
+                        } else {
+                            throw new RetryException("Exception in retry", lastThrowable);
```

Comment:
Do we need to wrap it?

Reply:
I get a Throwable from retryContext but I can only throw an Exception here, so I wrap it with RetryException.

Reply:
Maybe it is best to wrap the `Throwable` in a `RetryException` regardless and use the message, `Could not recover`.

Reply:
Maybe you are right